File: src/main/js/formulas/math/ACOTH.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ACOTH extends AbstractFormula {
  static SELF: ACOTH = new ACOTH();
  NAME = FormulaName.ACOTH;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (value <= 1 && value >= -1) {
      throw new NumException(
        `ACOTH parameter 1 value is ${value}. Valid values cannot be between -1 and 1 inclusive.`
      );
    }
    return 0.5 * Math.log((value + 1) / (value - 1));
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ACOTH.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NumException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ACOTH extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ACOTH;
  public static ACOTH SELF = new ACOTH();

  public ACOTH() {
    super();
  }

  public ACOTH(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    double value = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0])));
    if (value <= 1 && value >= -1) {
      throw new NumException(String.format("ACOTH parameter 1 value is %f. Valid values cannot be between -1 and 1 inclusive.", values));
    }
    return 0.5 * Math.log((value + 1) / (value - 1));
  }
}
